### PR TITLE
Bump python minimum supported version to 3.9

### PR DESCRIPTION
**Pull request by maflcko**
All supported operating systems ship with python 3.9 (or later), so bumping the minimum should not cause any issues. A bump will allow new code to use new python 3.9 features.

For reference:
* https://packages.debian.org/bullseye/python3
* https://packages.ubuntu.com/focal/python3.9
* FreeBSD 12/13 also ships with 3.9
* CentOS-like 8/9 also ships with 3.9 (and 3.11)
* OpenSuse Leap also ships with 3.9 (and 3.11) https://software.opensuse.org/package/python311-base

This is for Bitcoin Core 27.0 in 2024 (next year), not the soon upcoming 26.0 next month.

### CODE REVIEW

**Comment by Sjors** [comment]
Don't forget to bump `.python-version`

**Comment by luke-jr** [comment]
>CentOS Stream also ships with 3.9

Isn't Stream a rolling release? What about RHEL?

**Comment by Sjors** [comment]
I lightly tested fa5b76d1d8e1fb9bbb4a5cb11385c9b45bb91214 on Intel macOS 13.5 by installing Python 3.9.17 via pyenv, running `make deploy` as well as the functional tests. Seems fine.

**Review by jamesob — APPROVED** [review]
GH ACK https://github.com/bitcoin/bitcoin/pull/28211/commits/fa5b76d1d8e1fb9bbb4a5cb11385c9b45bb91214

If CI is happy, looks good to me.

**Comment by hebasto** [comment]
Concept ACK.

File: test/functional/test_framework/util.py [hunk 1]
```diff
@@ -287,10 +287,10 @@ def sha256sum_file(filename):
     return h.digest()
 
 
-# TODO: Remove and use random.randbytes(n) instead, available in Python 3.9
+# TODO: Remove and use random.randbytes(n) directly
```
Comment (hebasto) [inline test/functional/test_framework/util.py:290]: Why was this TODO comment left if it has been completed?

**Review by jamesob — APPROVED** [review]
ACK fa25e8b0a1610553014c786428f146ef9c694678 ([`jamesob/ackr/28211.1.MarcoFalke.bump_python_minimum_supp`](https://github.com/jamesob/bitcoin/tree/ackr/28211.1.MarcoFalke.bump_python_minimum_supp))

Ran new lint (+ build) command, ran functional tests.

File: .python-version [hunk 2]
```diff
@@ -1 +1 @@
-3.8.17
+3.9.17
```
Comment (Sjors) [inline .python-version:1]: Bump to 3.9.18 if you have to retouch. This time I tested with that patch version.

File: ci/test/00_setup_env_native_nowallet_libbitcoinkernel.sh [hunk 3]
```diff
@@ -7,9 +7,9 @@
 export LC_ALL=C.UTF-8
 
 export CONTAINER_NAME=ci_native_nowallet_libbitcoinkernel
-export CI_IMAGE_NAME_TAG="docker.io/ubuntu:20.04"
-# Use minimum supported python3.8 and clang-10, see doc/dependencies.md
-export PACKAGES="python3-zmq clang-10 llvm-10 libc++abi-10-dev libc++-10-dev"
-export DEP_OPTS="NO_WALLET=1 CC=clang-10 CXX='clang++-10 -stdlib=libc++'"
+export CI_IMAGE_NAME_TAG="docker.io/debian:bullseye"
+# Use minimum supported python3.9 and clang-10 (or best-effort clang-11), see doc/dependencies.md
+export PACKAGES="python3-zmq clang-11 llvm-11 libc++abi-11-dev libc++-11-dev"
```
Comment (Sjors) [inline ci/test/00_setup_env_native_nowallet_libbitcoinkernel.sh:12]: Shouldn't we keep clang-10 here as long as that's the minimum requirement?

**Review by Sjors — COMMENTED** [review]
Tested fa25e8b0a1610553014c786428f146ef9c694678 by running the functional tests on macOS 13.5.1 and the Docker linter on Ubuntu. Looks good except I don't understand the rationale for also bumping clang.

**Comment by Sjors** [comment]
ACK fa25e8b0a1610553014c786428f146ef9c694678

**Comment by fanquake** [comment]
Makes sense for this to go in first, and rebase the compiler bumps.

**Pull request merged**